sdk/core/core-paging/src/getPagedAsyncIterator.ts

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { PagedAsyncIterableIterator, PageSettings, PagedResult } from "./models";
+
+/**
+ * returns an async iterator that iterates over results. It also has a `byPage`
+ * method that returns pages of items at once.
+ *
+ * @param pagedResult - an object that specifies how to get pages.
+ * @returns a paged async iterator that iterates over results.
+ */
+export function getPagedAsyncIterator<TElement, TPage = TElement[], TPageSettings = PageSettings>(
+  pagedResult: PagedResult<TPage>
+): PagedAsyncIterableIterator<TElement, TPage, TPageSettings> {
+  const iter = getItemAsyncIterator<TElement, TPage>(pagedResult);
+  return {
+    next() {
+      return iter.next();
+    },
+    [Symbol.asyncIterator]() {
+      return this;
+    },
+    byPage:
+      pagedResult?.byPage ??
+      ((settings?: PageSettings) => {
+        return getPageAsyncIterator(pagedResult, settings?.maxPageSize);
+      })
+  };
+}
+
+async function* getItemAsyncIterator<TElement, TPage>(
+  pagedResult: PagedResult<TPage>,
+  maxPageSize?: number
+): AsyncIterableIterator<TElement> {
+  const metaInfo = { isArray: false };
+  const pages = getPageAsyncIterator<TPage>(pagedResult, maxPageSize, metaInfo);
+  const firstVal = await pages.next();
+  // if the result does not have an array shape, i.e. TPage = TElement, then we return it as is
+  if (!metaInfo.isArray) {
+    yield firstVal.value;
+    // `pages` is of type `AsyncIterableIterator<TPage>` but TPage = TElement in this case
+    yield* (pages as unknown) as AsyncIterableIterator<TElement>;
+  } else {
+    yield* firstVal.value;
+    for await (const page of pages) {
+      // pages is of type `AsyncIterableIterator<TPage>` so `page` is of type `TPage`. In this branch,
+      // it must be the case that `TPage = TElement[]`
+      yield* (page as unknown) as TElement[];
+    }
+  }
```

Comment:
I think we want to avoid `out` params, and use result objects/tuples instead where necessary, but I'm not convinced this is actually necessary for this usage. Can't we check if the yielded value is an array directly instead of relying on `metaInfo`? e.g.

```suggestion
  const pages = getPageAsyncIterator<TPage>(pagedResult, maxPageSize);
  const firstVal = await pages.next();
  // if the result does not have an array shape, i.e. TPage = TElement, then we return it as is
  if (!Array.isArray(firstVal.value)) {
    yield firstVal.value;
    // `pages` is of type `AsyncIterableIterator<TPage>` but TPage = TElement in this case
    yield* (pages as unknown) as AsyncIterableIterator<TElement>;
  } else {
    yield* firstVal.value;
    for await (const page of pages) {
      // pages is of type `AsyncIterableIterator<TPage>` so `page` is of type `TPage`. In this branch,
      // it must be the case that `TPage = TElement[]`
      yield* (page as unknown) as TElement[];
    }
  }
```

Reply:
Done.